Core/clim-basic/extended-streams/stream-input.lisp

```diff
@@ -20,57 +20,68 @@
-;;; this is the right place to do the transformation...
-
-;;;  Why exactly do we want to see #\Delete instead of #\Backspace?
-;;;  There is a separate Delete key, unless your keyboard is
-;;;  strange. --Hefner
-
-(defconstant +read-char-map+ '((#\Return . #\Newline)
-                               #+nil (#\Backspace . #\Delete)))
-
-(defvar *abort-gestures* '(:abort))
-
-(defvar *accelerator-gestures* nil)
+(in-package #:clim-internals)
 
 (defvar *input-wait-test* nil)
 (defvar *input-wait-handler* nil)
 (defvar *pointer-button-press-handler* nil)
 
-(define-condition abort-gesture (condition)
-  ((event :reader %abort-gesture-event :initarg :event)))
-
-(defmethod abort-gesture-event ((condition abort-gesture))
-  (%abort-gesture-event condition))
-
-(define-condition accelerator-gesture (condition)
-  ((event :reader %accelerator-gesture-event :initarg :event)
-   (numeric-argument :reader %accelerator-gesture-numeric-argument
-                     :initarg :numeric-argument
-                     :initform 1)))
-
-(defmethod accelerator-gesture-event ((condition accelerator-gesture))
-  (%accelerator-gesture-event condition))
-
-(defmethod accelerator-gesture-numeric-argument
-    ((condition accelerator-gesture))
-  (%accelerator-gesture-numeric-argument condition))
-
+;;; X11 returns #\Return where we want to see #\Newline at the stream-read-char
+;;; level.  Dunno if this is the right place to do the transformation... --moore
 (defun char-for-read (char)
-  (let ((new-char (cdr (assoc char +read-char-map+))))
-    (or new-char char)))
+  (check-type char (or character null))
+  (case char
+    (#\return #\newline)
+    (otherwise char)))
+
+(defun event-char (event)
+  (let ((modifiers (event-modifier-state event)))
+    (when (or (zerop modifiers)
+              (eql modifiers +shift-key+))
+      (keyboard-event-character event))))
+
+
+;;; Trampolines for stream-read-gesture and stream-unread-gesture.
+(defun read-gesture (&key (stream *standard-input*) timeout peek-p
+                       (input-wait-test *input-wait-test*)
+                       (input-wait-handler *input-wait-handler*)
+                       (pointer-button-press-handler *pointer-button-press-handler*))
+  (stream-read-gesture stream
+                       :timeout timeout
+                       :peek-p peek-p
+                       :input-wait-test input-wait-test
+                       :input-wait-handler input-wait-handler
+                       :pointer-button-press-handler pointer-button-press-handler))
 
-(defun unmap-char-for-read (char)
-  (let ((new-char (car (rassoc char +read-char-map+))))
-    (or new-char char)))
+(defun unread-gesture (gesture &key (stream *standard-input*))
+  (stream-unread-gesture stream gesture))
 
+
 (defclass input-kernel-mixin ()
   ((input-buffer :initarg :input-buffer :accessor stream-input-buffer :type vector)))
 
+;;; Intentionally not specialized to allow sheets.
```

Comment:
The spec says `stream` is an extended input stream.

Reply:
Yes, that's why I've mentioned in the comment, that we are intentionally not specializing it, to allow setting input to any sheet (it is a deliberate choice). It seems useful to me that we can give the input focus to a basic-sheet.